Remove security and privacy issue Gravatar #2202

Rotzbua opened the pull request:
Fixes: Remove security and privacy issue Gravatar #2202

Change: Replaces Gravatar image with an user icon.

w00fz commented:
I'm failing to see how this can be considered a privacy concern, in this context specifically. 
We are talking about a use of Gravatar that is purely visible to the user itself, within the admin, when logged in. This is not shared or visible anywhere else, outside your logged in user. If you can see this user Gravatar, you can already see the user email and change password.

I would be ok making this an opt-in solution, but I did checkout your PR and it's not quite ready for a merge. Here the issues I see:
![CleanShot 2022-02-04 at 10 07 51@2x](https://user-images.githubusercontent.com/11734/152580808-5aa9ae1b-c421-450f-aaee-2b64817aabb7.png)

1. This seems to be still using Gravatar
2. This should be consistent with the size of the avatar view, 200px font-size perhaps
3. If you want this to make it opt-in and disabled by default, we do need an option to change this behavior, right now I am not entirely sure what the PR is supposed to do but it looks like it's missing something.

Please make those changes and we can integrate this

froschdesign commented:
@w00fz 
> I'm failing to see how this can be considered a privacy concern, in this context specifically.

The Gravatar service itself is the problem. The reference was given in the related issue report: https://monitor.firefox.com/breach-details/Gravatar

> 3. If you want this to make it opt-in and disabled by default…

Regardless of security issues, the use of external services should **always** be optional.

Rotzbua commented on themes/grav/scss/template/_admin.scss:
```diff
@@ -1114,7 +1114,7 @@ body.sidebar-quickopen #admin-main {
     }
 
-    .gravatar {
+    avatar-upload-hint {
```
dot missing

Rotzbua commented on themes/grav/scss/preset.scss:
```diff
@@ -1286,7 +1286,7 @@ form {
     }
 
-    .gravatar {
+    avatar-upload-hint {
```
dot missing

Rotzbua commented:
>     1. This seems to be still using Gravatar

Left avatar template is:
```twig
{% set user_avatar = admin.user.getAvatarUrl() %}
{% if user_avatar %}
```
Right avatar template is:
```twig
{% if data.avatar %}
    <label><img src="{{ data.getAvatarUrl() }}" /></label>
```

Indeed this is a bug in the plugin. The left template does it correctly. The right template is wrong, because it only checks for local avatar by `data.avatar`. But grav also allowes external avatar which should be requested by `getAvatarUrl` as documented in [core](https://github.com/getgrav/grav/blob/163cd5a03f0d9713a3b8a529dd418d44352f2614/system/src/Grav/Common/User/Traits/UserTrait.php#L123) :

>     * Note: if there's no local avatar image for the user, you should call getAvatarUrl() to get the external avatar URL.


So you are right that both templates should behave similar. I changed them in the PR so they just relay on `getAvatarUrl`.

>     2. This should be consistent with the size of the avatar view, 200px font-size perhaps

I removed the fallback. A non existing avatar does not destroy the layout. If there should be a default image it should be served by core functions.

>     3. If you want this to make it opt-in and disabled by default, we do need an option to change this behavior, right now I am not entirely sure what the PR is supposed to do but it looks like it's missing something.

I reworked the PR. Now it does not change the behavour, because `getAvatarUrl` serves by default a fallback avatar from gravatar. So currently there should be no beak until `getAvatarUrl` is changed.

mahagr commented:
I prefer this one:

https://github.com/getgrav/grav/pull/3550

rhukster commented:
See: https://github.com/getgrav/grav/pull/3551
